### DISCUSSION TITLE
Failed Experiment: Cut Cross Entropy reduces VRAM by 2x but loss with base still better

### DISCUSSION THREAD

**Discussion by omkaark** [karpathy/nanochat#79] [General]
The only deviation of setup is I train on 1xA100, depth=10 and only 4k steps vs the original repo. 

I hypothesized that CCE w/ Kahan Sum allows 2x batch size and approximately produces the same results as torch's cross_entropy so it should produce a better loss curve. I get an increase in tok/s trained, lower MFU but loss curve for base is better. 

<img width="2864" height="1968" alt="image" src="https://github.com/user-attachments/assets/f41b099e-f408-41be-8b5c-13dc86c3dcf0" />

**Comment by karpathy** [comment]
Ty for posting. I have a note in the code in gpt.py:

```python
            # training mode: compute and return the loss
            # TODO: experiment with Liger Kernels / chunked cross-entropy etc.
            logits = self.lm_head(x)
            logits = softcap * torch.tanh(logits / softcap) # logits softcap
            logits = logits.float() # use tf32/fp32 for logits
            loss = F.cross_entropy(logits.view(-1, logits.size(-1)), targets.view(-1), ignore_index=-1, reduction=loss_reduction)
            return loss
```

my understanding is this is basically the same idea of chunked cross entropy. Possibly this is a thing we reach for once we're training models so large that we're down all the way to device_batch_size=1, and then this makes sense to reach for to make things fit (?). But yes I think I expected a bit more still.
